README.md
# Styled components media query
[![npm (scoped with tag)](https://img.shields.io/npm/v/styled-components-media-query/latest.svg?style=flat-square)](https://npmjs.com/package/styled-components-media-query)
[![npm](https://img.shields.io/npm/dt/styled-components-media-query.svg?style=flat-square)](https://npmjs.com/package/styled-components-media-query)

> Simple styled components utility that you can use for **building media queries**.

## Install
- yarn: `yarn add styled-components-media-query`
- npm: `npm i --save styled-components-media-query`

## Usage

```js
import { styled, css } from 'styled-components'
import MqInit from 'styled-components-media-query'

// initialize
const bp = {s: 400, sl: 500, m: 768, ml: 992, l: 1100}
const mq = MqInit({ bp })

// use like this
const ComponentStyled = styled.div`
 background: red;
 
 // max-width(399px)
 ${mq(null, 's')(css`
   background: green;
 `)}
 
 // between min-width(400px) and max-width(767px)
 ${mq('s', 'm')(css`
   background: black;
 `)}
 
 // min-width(768px)
 ${mq('m')(css`
   background: orange;
 `)}
 
 // min-width(1100px)
 ${mq('l')(css`
   background: purple;    
 `)}
`
```

```css
/* compiled css */
.myClass {
  background: red;
}
@media only screen and (max-width: 399px) {
  .myClass {
    background: green;
  }
}
@media only screen and (min-width: 400px) and (max-width: 767px) {
  .myClass {
    background: black;
  }
}
@media only screen and (min-width: 768px) {
  .myClass {
    background: orange;
  }
}
@media only screen and (min-width: 1100px) {
  .myClass {
    background: purple;
  }
}
```

## Api

---

<details><summary><code>MqInit({ bp, type })</code></summary><p>

- `bp` (object): a breakpoint object (optional).
- `type` (string): can be 'width' or 'height' (optional).

```js
import { styled, css } from 'styled-components'
import MqInit from 'styled-components-media-query'

const bp = {s: 300, m: 500, l: 800}

// this will build min-width, max-width queries
const mqWidth = MqInit({ bp, type: 'width' })

// this will build min-height, max-height queries
const mqHeight = MqInit({ bp, type: 'height' })
```

If you don't pass any value when initializing `MqInit`, it will use the default `bp` object and `type` string, which are:

```js
const bp = {  
  xxxs: 0,
  xxs: 200,
  xs: 320,
  s: 400,
  sl: 500,
  m: 768,
  ml: 992,
  l: 1100,
  xl: 1200,
  xxl: 1300,
  xxxl: 1400,
  xxxxl: 1500,
  xxxxxl: 1600,
  xxxxxxl: 1700,
}
const type = 'width'
```

You can also use `mq` directly without initializing `MqInit` at all, but **I don't recommend using my breakpoints**, always better to tailor them for the specific project.
 
```js
import { styled, css } from 'styled-components'
import {mq} from 'styled-components-media-query'

const CompStyled = styled.div`
  // min-width(768px)
  ${mq('m')(css``)} 
`
```

</p></details>

---

<details><summary><code>mq(min, max)</code></summary><p>

- `min` (string|number|null): min-width or min-height value.
- `max` (string|number|null): max-width or max-height value.

You can use the mq function to specify min-width, max-width, min-height and max-height. Examples:

```js
import { styled, css } from 'styled-components'
import MqInit from 'styled-components-media-query'

const bp = {s: 300, m: 500, l: 800}
const mq = MqInit({ bp })
const mqHeight = MqInit({ bp, type: 'height' })

const CompStyled = styled.div`
  // min-width(500px)
  ${mq('m')(css``)}         
  
  // max-width(500px)
  ${mq(null, 'm')(css``)}
  
  // between min-width(500px) and max-width(799px)
  ${mq('m', 'l')(css``)}
  
  // min-height(500px)
  ${mqHeight('m')(css``)}         
  
  // max-height(500px)
  ${mqHeight(null, 'm')(css``)}
  
  // between min-height(500px) and max-height(799px)
  ${mqHeight('m', 'l')(css``)}
`
```

If you specify a string that is not in the `bp` object, it will throw an error, however, if needed, **you can pass a specific number directly**:

```js
import { styled, css } from 'styled-components'
import MqInit from 'styled-components-media-query'

const mq = MqInit()

const CompStyled = styled.div`
  // min-width(300px)
  ${mq(300)(css``)}         
  
  // max-width(603px)
  ${mq(null, 603)(css``)}
  
  // between min-width(23px) and max-width(900px)
  ${mq(23, 900)(css``)}
`
```

Also, the `css` helper is not necessary, so if you want you can use `mq` like this:

```js
import { styled } from 'styled-components'
import MqInit from 'styled-components-media-query'

const mq = MqInit()

const CompStyled = styled.div`
  // min-width(300px)
  ${mq(300)(``)} // note that 'css' helper is missing
`
```
</p></details>

---

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.15"></a>
## [1.0.15](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.14...v1.0.15) (2019-01-10)



<a name="1.0.14"></a>
## [1.0.14](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.13...v1.0.14) (2019-01-10)



<a name="1.0.13"></a>
## [1.0.13](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.12...v1.0.13) (2019-01-04)



<a name="1.0.12"></a>
## [1.0.12](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.11...v1.0.12) (2019-01-04)



<a name="1.0.11"></a>
## [1.0.11](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.10...v1.0.11) (2019-01-04)



<a name="1.0.10"></a>
## [1.0.10](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.9...v1.0.10) (2019-01-04)



<a name="1.0.9"></a>
## [1.0.9](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.8...v1.0.9) (2019-01-04)



<a name="1.0.8"></a>
## [1.0.8](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.7...v1.0.8) (2019-01-04)



<a name="1.0.7"></a>
## [1.0.7](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.6...v1.0.7) (2019-01-04)



<a name="1.0.6"></a>
## [1.0.6](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.5...v1.0.6) (2019-01-03)



<a name="1.0.5"></a>
## [1.0.5](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.4...v1.0.5) (2019-01-03)



<a name="1.0.4"></a>
## [1.0.4](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.3...v1.0.4) (2019-01-03)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/microcipcip/styled-components-media-query/compare/v1.0.2...v1.0.3) (2019-01-03)



<a name="1.0.2"></a>
## [1.0.2](https://github.com/microcipcip/styled-components-mq/compare/v1.0.1...v1.0.2) (2019-01-03)


### Bug Fixes

* Fixing errors according to tslint.json ([2f58273](https://github.com/microcipcip/styled-components-mq/commit/2f58273))



<a name="1.0.1"></a>
## 1.0.1 (2019-01-03)


src/lib/core.spec.ts
// tslint:disable:no-expression-statement
// import test from 'ava'
// import { getBpValue, bpDefaults } from './core'
//
// test('Check that getBpValue returns the expected value', t => {
//   t.is(getBpValue(0, bpDefaults), 0)
//   t.is(getBpValue(10, bpDefaults), 10)
//   t.is(getBpValue(null, bpDefaults), -1)
//   t.is(getBpValue('m', bpDefaults), bpDefaults['m'])
// })


assets/js/search.js
var typedoc = typedoc || {};
            typedoc.search = typedoc.search || {};
            typedoc.search.data = {"kinds":{"32":"Variable","64":"Function","256":"Interface","2097152":"Object literal"},"rows":[{"id":0,"kind":256,"name":"Ibp","url":"interfaces/ibp.html","classes":"tsd-kind-interface tsd-is-not-exported"},{"id":1,"kind":2097152,"name":"bpDefaults","url":"globals.html#bpdefaults","classes":"tsd-kind-object-literal"},{"id":2,"kind":32,"name":"xxxs","url":"globals.html#bpdefaults.xxxs","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":3,"kind":32,"name":"xxs","url":"globals.html#bpdefaults.xxs","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":4,"kind":32,"name":"xs","url":"globals.html#bpdefaults.xs","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":5,"kind":32,"name":"s","url":"globals.html#bpdefaults.s","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":6,"kind":32,"name":"sl","url":"globals.html#bpdefaults.sl","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":7,"kind":32,"name":"m","url":"globals.html#bpdefaults.m","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":8,"kind":32,"name":"ml","url":"globals.html#bpdefaults.ml","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":9,"kind":32,"name":"l","url":"globals.html#bpdefaults.l","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":10,"kind":32,"name":"xl","url":"globals.html#bpdefaults.xl","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":11,"kind":32,"name":"xxl","url":"globals.html#bpdefaults.xxl","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":12,"kind":32,"name":"xxxl","url":"globals.html#bpdefaults.xxxl","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":13,"kind":32,"name":"xxxxl","url":"globals.html#bpdefaults.xxxxl","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":14,"kind":32,"name":"xxxxxl","url":"globals.html#bpdefaults.xxxxxl","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":15,"kind":32,"name":"xxxxxxl","url":"globals.html#bpdefaults.xxxxxxl","classes":"tsd-kind-variable tsd-parent-kind-object-literal","parent":"bpDefaults"},{"id":16,"kind":64,"name":"getBpValue","url":"globals.html#getbpvalue","classes":"tsd-kind-function"},{"id":17,"kind":64,"name":"MqInit","url":"globals.html#mqinit","classes":"tsd-kind-function"},{"id":18,"kind":32,"name":"mqExec","url":"globals.html#mqexec","classes":"tsd-kind-variable"}]};

src/lib/core.ts
import { css } from 'styled-components'

export const bpDefaults = {
  xxxs: 0,
  xxs: 200,
  xs: 320,
  s: 400,
  sl: 500,
  m: 768,
  ml: 992,
  l: 1100,
  xl: 1200,
  xxl: 1300,
  xxxl: 1400,
  xxxxl: 1500,
  xxxxxl: 1600,
  xxxxxxl: 1700,
}

interface Ibp {
  readonly [bp: string]: number
}

export const getBpValue = (
  val: number | string | null | undefined,
  bp: Ibp
): number => {
  switch (typeof val) {
    case 'number':
      return val
    case 'string':
      if (!bp.hasOwnProperty(val)) {
        throw new Error(
          `Media Query error: oops, you passed a value that is not defined in the 'bp' object.`
        )
      }
      if (typeof bp[val] !== 'number') {
        throw new Error(
          `Media Query error: oops, you passed a value that is not a proper number.`
        )
      }

      return bp[val]
    default:
      return -1
  }
}

/**
 * Initialize media query
 *
 * ### Usage
 * ```js
 * import { styled, css } from 'styled-components'
 * import MqInit from 'styled-components-media-query'
 *
 * // initialize
 * const bp = {s: 400, sl: 500, m: 768, ml: 992, l: 1100}
 * const mq = MqInit({ bp })
 *
 * // use like this
 * const ComponentStyled = styled.div`
 *  position: relative;
 *  ${mq('m')(css`
 *    position: absolute;
 *  `)}
 *  ${mq('l')(css`
 *    position: fixed;
 *  `)}
 * `
 * ```
 *
 * @param bp      Pass a breakpoint object to customize the breakpoints you wish to use.
 * @param type    Pass a type string to either use min/max-width or min/max-height.
 * @returns       A styled-component css string.
 */

const MqInit = ({ bp = bpDefaults, type = 'width' } = {}) => (
  min?: string | number | null,
  max?: string | number | null
) => (contentCSS: any) => {
  if (!Array.isArray(contentCSS) && typeof contentCSS !== 'string') {
    throw new Error(
      `Media Query error: oops, you passed a invalid argument. Valid arguments are 'string' or the styled components 'css' helper.`
    )
  }

  const minV = getBpValue(min, bp)
  const maxV = getBpValue(max, bp)
  const content = Array.isArray(contentCSS) ? contentCSS : css`${contentCSS}`

  if (minV >= 0 && maxV === -1) {
    return css`
      @media only screen and (min-${type}: ${minV}px) {
        ${content}
      }
    `
  } else if (minV === -1 && maxV >= 0) {
    return css`
      @media only screen and (max-${type}: ${maxV - 1}px) {
        ${content}
      }
    `
  } else if (minV >= 0 && maxV >= 0) {
    return css`
      @media only screen and (min-${type}: ${minV}px) and (max-${type}: ${maxV -
      1}px) {
        ${content}
      }
    `
  }
  return css``
}

export const mqExec = MqInit()

export default MqInit
